packages/openapi3/src/openapi.ts

```diff
@@ -1097,13 +1097,49 @@ function createOAPIEmitter(program: Program, options: ResolvedOpenAPI3EmitterOpt
     );
   }
 
+  function checkAdditionalProperties(program: Program, model: Model): boolean {
+    const baseModel = model.baseModel;
+    if (!baseModel || baseModel.name !== "Record") {
+      return false;
+    }
+    const templateArg = baseModel.templateMapper?.args.at(0);
+    if (!templateArg) {
+      return false;
+    }
+    switch (templateArg.kind) {
+      case "Intrinsic":
+        return templateArg.name === "unknown";
+      case "Scalar":
+        let isAssignable = true;
+        for (const [_, prop] of model.properties) {
+          const [isTypeAssignable, diagnostics] = program.checker.isTypeAssignableTo(
+            prop.type,
+            templateArg,
+            prop
+          );
+          for (const diag of diagnostics) {
+            program.reportDiagnostic(diag);
+          }
+          isAssignable = isAssignable && isTypeAssignable;
+        }
+        return isAssignable;
+      default:
+        return false;
+    }
+  }
+
   function getSchemaForModel(model: Model, visibility: Visibility) {
     let modelSchema: OpenAPI3Schema & Required<Pick<OpenAPI3Schema, "properties">> = {
       type: "object",
       properties: {},
       description: getDoc(program, model),
     };
 
+    const hasAdditionalProperties = checkAdditionalProperties(program, model);
+    if (hasAdditionalProperties) {
+      modelSchema.additionalProperties = true;
```

Comment:
We only want additionalProperties = true when the model extends `Record<unknown>`.  if it extends `Record<anything-but-unknown>` then we want to leave additionalProperties representation in OpenAPI3 as is. (which is, setting it to the schema for the type in the first template argument.

Reply:
Thanks for the clarification. Please take a look at the updated test cases.

Reply:
> then we want to leave additionalProperties representation in OpenAPI3 as is. (which is, setting it to the schema for the type in the first template argument.

Unclear if you mean "don't change anything" or set `additionalProperties: <schema type>`?

Reply:
The PR now reflects the test cases we discussed. Please take another look.